### DISCUSSION TITLE
Override improvements

### DISCUSSION THREAD

**Discussion by appelgriebsch** [LazyVim/LazyVim#22] [Ideas]
I just made big steps in my migration from a custom Neovim setup to a configuration based on the starter template for LazyVim (https://github.com/appelgriebsch/Nv).

During the work I figured that it would be nice to add some 'override' functionality to the following items:

* cmp key mappings (in coding.lua)
* cmp sources (in coding.lua)
* additional cmp options (like borders)

* telescope key mappings (in editor.lua)
* telescope extension and their options (in editor.lua)

* mason ui options (in lsp.lua)
* alpha dashboard buttons (in ui.lua)

Maybe a similar override function as in the lualine configuration will be handy to provide a way to override some of the settings in those scenarios?!

**Comment by appelgriebsch** [comment]
Also: how do I create descriptions for new which-key sub-menus without having to overwrite the whole which-key configuration?...

Here the new DAP menu in wk:
![image](https://user-images.githubusercontent.com/6803419/211168229-3d0b052c-82c7-484a-9576-9c19bb2f4762.png)

**Comment by folke** [comment]
I'm working on something that would make configuring plugins easier, but for now:
- cmp: you'll have to override the whole config function in your personal spec, so just plugin name and config function. The rest can stay the same.
- dependencies, keys, event, ft, cmd props are merged, so if you want to add extra keys, you can just specify them in your local spec
- you can also get the `super` spec that you are overriding in your own spec. See my telescope config as an example:
- https://github.com/folke/dot/blob/master/config/nvim/lua/plugins/telescope.lua
- mason & alpha: for now you need to override the whole config function

for which-key, you can do something like the below:

```lua
{
  "folke/which-key.nvim",
  config = function(plugin)
    -- run original config
    plugin._.super.config()
   require("which-key").register(add your own stuff here)
```

I hope this helps!

**Comment by folke** [comment]
I just released the new functionality in `lazy.nvim` that makes extending much easier.

See some examples from my config:

* cmp: https://github.com/folke/dot/blob/a93b0ca30aa8739f3c08800cf0214cd8d2fbbd80/config/nvim/lua/plugins/coding.lua#L179
* telescope: https://github.com/folke/dot/blob/master/config/nvim/lua/plugins/telescope.lua
* for mason, just add extra config in the `opts` property
* for dashboard, you can make `opts` a function with params `opts = function(plugin, opts)`. That opts param will have the dashboard that you can then customize.

for which-key:

```lua
{
  "folke/which-key.nvim",
  opts = function()
   require("which-key").register(add your own stuff here)
  end
}
```

I removed all the special properties like `override`, `ensure_installed`, `servers`, `setup_server`. This is now all handled by the `opts` property.

For lsp check here https://github.com/LazyVim/LazyVim/blob/2135bc144c0ea18d550d40c38125e5cd73965c73/lua/lazyvim/plugins/lsp/init.lua#L14

`props` are always merged with spec parents. When it is a function, the parent opts are passed as a parameter. You can then change the opts table directly, or return a whole new opts table that will be used instead.

Sorry again for breaking LazyVim, but I hope this will make it all a lot easier to configure and extend!

**Reply by appelgriebsch** [reply]
Hmmm looks like the override for mason - ensure_installed is not working if I split it up into dedicated language-modules.... https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/extras/lang/jdtls.lua#L17

Import it here: https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/lsp.lua#L5, but it doesn't install the mason packages. That would be a big plus as for some language modules we may need more than one mason package (e.g. java lsp, test and debug package)

**Reply by folke** [reply]
That's because I use `vim.tbl_deep_extend`. That only merges tables, not lists.

You should do something similar as what I did here for typescript and treesitter:
https://github.com/LazyVim/LazyVim/blob/f7475ab7aadcfe89b613490c7f2169469130c767/lua/lazyvim/plugins/extras/lang/typescript.lua#L4

**Reply by appelgriebsch** [reply]
:facepalm:

**Reply by folke** [reply]
Your config looks really cool!

https://github.com/appelgriebsch/Nv/blob/85eec65b05bae9f07e9f9bd2e5116d539031b44a/lua/plugins/extras/lang/rust-analyzer.lua#L8

There you should use `vim.list_extend` though. That won't work what you have right now :)

Same in some other files

**Reply by appelgriebsch** [reply]
sorry, neovim lua is still new to me.... but thanks a lot for the feedback!

**Reply by appelgriebsch** [reply]
Hmmm I followed the changes you mentioned, and also replaced my custom jsonls with your new extra module. But when cleaning my .local/share/nvim files to start from scratch, it still don't load the Treesitter and Mason modules mentioned in the plugin spec opts override... 

I also tried to add some print(vim.inspect(opts)) to those override functions, but it doesn't seem to call them....

**Reply by folke** [reply]
What module doesn't work for you as an example? And where did you put the print statement that doesn't work?

**Reply by folke** [reply]
@appelgriebsch  You are overriding treesitter completely in https://github.com/appelgriebsch/Nv/blob/main/lua/plugins/treesitter.lua

Specs are loaded/merged in order they are specified, so 
https://github.com/appelgriebsch/Nv/blob/4ff6e19d95d28fedddbef6e5055a88329dab4392/lua/config/lazy.lua#L18

Would override your extras.

You should probably move the import of the extras **after** your `import = "plugins"`

**Reply by folke** [reply]
You also don't need to add the lsp servers to ensure_installed for mason, since `mason-lspconfig` already takes care of that. You can fo course keep those additional tools in there. Just remove the lsp ones. (doesn't harm anything, just not needed)

**Reply by appelgriebsch** [reply]
Updated my configuration this morning and ran a bunch of tests especially with rust and java… works perfectly! 

Thanks a lot!

**Comment by folke** [comment]
See also https://github.com/folke/lazy.nvim#%EF%B8%8F-importing-specs-config--opts

**Comment by scottames** [comment]
This is all super helpful and awesome. Thanks for the awesome hard work @folke! Super excited to dig into Lazy more.
